lib/fast_jsonapi/serialization_core.rb

```diff
@@ -94,58 +94,47 @@ def id_from_record(record, params)
         record.id
       end
 
-      def parse_include_item(include_item)
-        return [include_item.to_sym] unless include_item.to_s.include?('.')
-
-        include_item.to_s.split('.').map!(&:to_sym)
-      end
-
-      def remaining_items(items)
-        return unless items.size > 1
-
-        [items[1..-1].join('.').to_sym]
+      def parse_includes_list(includes_list)
+        includes_list.each_with_object({}) do |include_item, include_sets|
+          include_base, include_remainder = include_item.to_s.split('.', 2)
```

Comment:
Why do we take just the first two split values here?

Reply:
The idea here is that if we're given a long/deep list of includes, like

```ruby
  BookSerializer.new(
    DATA.sample,
    include: [
      'authors',
      'authors.books',
      'authors.books.genre',
      'authors.books.genre.books',
      'authors.books.genre.books.authors',
      'authors.books.genre.books.genre',
      'genre',
      'genre.books'
    ]
  ).serializable_hash
```

The current serializer (BookSerializer, in this case), only cares about the unique, root inclusions requested. In the example above, that would only be `authors`. Everything after authors is outside of the BookSerializer's scope.

This is the bug in the current implementation. Partly, because it chops the include up into a flat array of all nested includes, it asks the BookSerializer to serialize "authors", then "books", then "genre", then "books", then "authors", then "genre", then "genre", then "books." The serialization attempts for BookSerializer relationships exponentially grows with the depth of those includes.

So, back to this implementation: it chops out the root association (first part) from each entry. It uniques it (because of how Hash keys work) and collects all of the "include_remainder" values to hand off to the next appropriate serializer. This method will turn that `include` array into a Hash that looks like:

```ruby
{
  'authors' => Set.new([
    'books',
    'books.genre',
    'books.genre.books',
    'books.genre.books.authors',
    'books.genre.books.genre'
  ]),
  'genre' => Set.new(['books'])
}
```

Because the BookSerializer only cares about the root associations included, it only needs the first segment of each include ("authors" and "genre") and it doesn't need to waste cycles parsing the rest of it. That will be the job of the next serializer in line. For example, the GenreSerializer would only see that it needs to include `["books"]`, while the AuthorsSerializer will get the Set of 5 entries.

Further, this implementation uses `Set`. This gives the benefit of ensuring the list of associations to include are unique. So, if this list is long and the caller mistakenly asked for "books.genre" twice, that relationship will only be iterated (to be checked and serialized) once, because pushing a duplicate entry into a Set is ignored.

Reply:
Alright, this makes sense. It wasn't clear since we do have further calls to `parse_includes_list`. Do you mind adding a little docstring maybe for the method? If not I can just merge it and update it...

Thanks! :bow:

Reply:
Added it here 1ce4677

Thanks again @nbibler for all the work here! :bow:

Reply:
Thanks for doing that! I've been swamped this week.